daffodil-core/src/main/scala/org/apache/daffodil/core/dpath/Expression.scala

```diff
@@ -773,6 +773,15 @@ case class RelativePathExpression(stepsRaw: List[StepExpression], isEvaluatedAbo
 sealed abstract class StepExpression(val step: String, val pred: Option[PredicateExpression])
   extends Expression {
 
+  def checkIfNodeIndexedLikeArray(): Unit = {
+    if (pred.isDefined) {
+      SDE(
+        "Expression contains indexing on a node element: %s.",
```

Comment:
I might recommend "non-array" instead of "node". I think we use that language in a few other places.